src/main/java/com/example/gymApp/service/TrainerService.java

```diff
@@ -1,131 +1,167 @@
+
+
 @Slf4j
 @Service
 public class TrainerService {
 
-  private final TrainerDAO trainerDAO;
+  private final TrainerRepository trainerRepository;
+  private final UserRepository userRepository;
+  private final TrainingTypeRepository trainingTypeRepository;
+  private final TraineeRepository traineeRepository;
+  private final TrainingRepository trainingRepository;
 
-  @Autowired
-  public TrainerService(TrainerDAO trainerDAO) {
-    this.trainerDAO = trainerDAO;
-  }
 
-  public boolean createTrainer(Long id, String firstName, String lastName, String specialization) {
-    try {
-      Trainer trainer = new Trainer();
-      trainer.setId(id);
-      trainer.setFirstName(firstName);
-      trainer.setLastName(lastName);
-      trainer.setSpecialization(specialization);
-
-      validateTrainer(trainer);
-
-      if (trainer.getUsername() == null || trainer.getUsername().isEmpty()) {
-        String generatedUsername = ProfileService.generateUsername(firstName, lastName);
-        trainer.setUsername(generatedUsername);
-        log.info("Generated username for trainer {}: {}", id, generatedUsername);
-      }
-      try {
-        if (trainer.getPassword() == null || trainer.getPassword().isEmpty()) {
-          String generatedPassword = ProfileService.generateRandomPassword();
-          trainer.setPassword(generatedPassword);
-          log.info("Generated password for trainer {}: {}", id, generatedPassword);
-        }
-      } catch (Exception e) {
-        log.error("Error generating password for trainer {}: {}", id, e.getMessage());
-        throw new RuntimeException("Failed to generate password for trainer.");
-      }
-
-
-      if (!trainerDAO.createTrainer(trainer)) {
-        throw new NoSuchElementException("Trainer with ID " + id + " could not be created.");
-      }
-      log.info("Trainer {} has been created successfully.", id);
-      return true;
-
-    } catch (IllegalArgumentException e) {
-      log.error("Error creating trainer: {}", e.getMessage());
-      return false;
-    }
+  @Autowired
+  public TrainerService(TrainerRepository trainerRepository, UserRepository userRepository,
+      TrainingTypeRepository trainingTypeRepository, TraineeRepository traineeRepository,
+      TrainingRepository trainingRepository) {
+    this.trainerRepository = trainerRepository;
+    this.userRepository = userRepository;
+    this.trainingTypeRepository = trainingTypeRepository;
+    this.traineeRepository = traineeRepository;
+    this.trainingRepository = trainingRepository;
   }
 
-  public boolean updateTrainer(Long id, String firstName, String lastName, String specialization) {
-    try {
+  @Transactional
+  public Trainer updateTrainer(Trainer trainer, String name, String lastName, String username,
+      String password, Boolean activeStatus, String specialization) {
 
-      Trainer trainer = getTrainerById(id);
+    Optional<TrainingType> trainingTypeOptional = trainingTypeRepository.findByName(specialization);
```

Comment:
You can create a separate method in the service layer where you get the trainingType and do all the checking for it. Right now you have duplicates of this part of the code:
```
Optional<TrainingType> trainingType = trainingTypeRepository.findByName(specialization);
    if (trainingType.isEmpty()) {
      throw new IllegalArgumentException("Specialization not found in the database");
    }
```